# **Predicting Kilocalories Based on Nutritional Information**
*Exploratory Data Analysis* | *Feature Selection* | *Model Building* | *Model Evaluation* | *Regression* | *AutoML* | *SHAP Values*

* Author: Vishnu Rao-Sharma
* Published: Fall 2023

## Abstract
With this dataset, I intend to find the relationship between macro/micro nutrients and Kilocalories in food. I will use SHRUTI SAXENA's [Food Nutrition Dataset](https://www.kaggle.com/datasets/shrutisaxena/food-nutrition-dataset?resource=download&select=food.csv) which is based on the USDA's [FoodData Central database](https://fdc.nal.usda.gov/fdc-app.html#/) to build a models that can predict the calories in a food based on key nutritional information. 

To begin, I preprocess the raw dataset and analyze the incoming dataset. This analysis helped determine the most important for predicting calories:
* 

Using these selected features, I build a few models with the selected features using the following regression techniques:
* OLS Linear Regression
* Gradient Boosting Regression (GBM)
* XGBoost Regression (XGB)

Before evaluating the models, I perform a grid search with k=5 cross-validation to find solid hyperparameters for the non-linear regression models. 

Finally, I compare model accuracy on the test set based on RMSE, and compare interpretability with feature SHAP values. This evaluation finds that the model built using UPDATE had the lowest RMSE. 

## Preparing Libraries and Setting-up AutoML
Before we get into the dataset, let's start by getting our packages, libraries, and Auto-ML config in order

In [1]:
# Installing necessary pkgs
!pip install --upgrade pip
!pip install requests
!pip install tabulate
!pip install "colorama>=0.3.8"
!pip install future
!pip install shap
!pip install xgboost
!pip install gbm
!pip3 install -f http://h2o-release.s3.amazonaws.com/h2o/latest_stable_Py.html h2o

Looking in links: http://h2o-release.s3.amazonaws.com/h2o/latest_stable_Py.html


In [2]:
# Importing libs
import h2o
from h2o.automl import H2OAutoML
import random, os, sys
import psutil
import random
import logging
import pandas as pd
import seaborn as sns
import numpy as np
from matplotlib import pyplot as plt
import statsmodels.api as sm
from scipy.stats import shapiro
from statistics import LinearRegression
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split
from mlxtend.feature_selection import SequentialFeatureSelector as SFS
from sklearn.feature_selection import VarianceThreshold
from mlxtend.plotting import plot_sequential_feature_selection as plot_sfs
from mlxtend.evaluate import bias_variance_decomp

In [3]:
# Setting up H2O VM
min_mem_size=6 
pct_memory=0.5
virtual_memory=psutil.virtual_memory()
min_mem_size=int(round(int(pct_memory*virtual_memory.available)/1073741824,0))
print(min_mem_size)

port_no=random.randint(5555,55555)
try:
  h2o.init(strict_version_check=False,min_mem_size_GB=min_mem_size,port=port_no) # start h2o
except:
  logging.critical('h2o.init')
  h2o.download_all_logs(dirname=logs_path, filename=logfile)    
  h2o.cluster().shutdown()
  sys.exit(2)

1
Checking whether there is an H2O instance running at http://localhost:48164..... not found.
Attempting to start a local H2O server...
  Java Version: openjdk version "17.0.4" 2022-07-19; OpenJDK Runtime Environment Temurin-17.0.4+8 (build 17.0.4+8); OpenJDK 64-Bit Server VM Temurin-17.0.4+8 (build 17.0.4+8, mixed mode)
  Starting server from /Users/vraosharma/Desktop/Python/DSEMT/exploratory-data-analysis/venv/lib/python3.10/site-packages/h2o/backend/bin/h2o.jar
  Ice root: /var/folders/36/syxnttfx02z8nxmjh2_llhnr0000gn/T/tmp5ezcsstb
  JVM stdout: /var/folders/36/syxnttfx02z8nxmjh2_llhnr0000gn/T/tmp5ezcsstb/h2o_vraosharma_started_from_python.out
  JVM stderr: /var/folders/36/syxnttfx02z8nxmjh2_llhnr0000gn/T/tmp5ezcsstb/h2o_vraosharma_started_from_python.err
  Server is running at http://127.0.0.1:48164
Connecting to H2O server at http://127.0.0.1:48164 ... successful.


--------------------------  ---------------------------------
H2O_cluster_uptime:         03 secs
H2O_cluster_timezone:       America/New_York
H2O_data_parsing_timezone:  UTC
H2O_cluster_version:        3.44.0.1
H2O_cluster_version_age:    1 month and 24 days
H2O_cluster_name:           H2O_from_python_vraosharma_ykpc1h
H2O_cluster_total_nodes:    1
H2O_cluster_free_memory:    4 Gb
H2O_cluster_total_cores:    8
H2O_cluster_allowed_cores:  8
H2O_cluster_status:         locked, healthy
H2O_connection_url:         http://127.0.0.1:48164
H2O_connection_proxy:       {"http": null, "https": null}
H2O_internal_security:      False
Python_version:             3.10.6 final
--------------------------  ---------------------------------

## Importing and Pre-processing the Dataset
In this section we will evaluate the hygiene of the dataset and clean it up as necessary.

In [4]:
# Addding the dataset from github https://github.com/vraosharma-northeastern/exploratory-data-analysis/blob/main/Nutrition%20/food.csv
!wget https://raw.githubusercontent.com/vishnuraosharma/exploratory-data-analysis/main/Kcal%20Predictions/food.csv

zsh:1: command not found: wget


In [5]:
#Reading the file into a dataframe and viewing the first few rows
df = pd.read_csv('food.csv')

# Loop through columns and remove redundant 'Data.' tag from column names
for col in df.columns:
    new_col = col.replace('Data.', '')  # Remove 'Data.' from the column name
    df.rename(columns={col: new_col}, inplace=True)  # Rename the column name in dataframe

# Show the first few rows of the dataset
df.head()

Category               Description  Nutrient Data Bank Number  \
0      BUTTER          BUTTER,WITH SALT                       1001   
1      BUTTER  BUTTER,WHIPPED,WITH SALT                       1002   
2  BUTTER OIL      BUTTER OIL,ANHYDROUS                       1003   
3      CHEESE               CHEESE,BLUE                       1004   
4      CHEESE              CHEESE,BRICK                       1005   

   Alpha Carotene   Ash  Beta Carotene  Beta Cryptoxanthin  Carbohydrate  \
0               0  2.11            158                   0          0.06   
1               0  2.11            158                   0          0.06   
2               0  0.00            193                   0          0.00   
3               0  5.11             74                   0          2.34   
4               0  3.18             76                   0          2.79   

   Cholesterol  Choline  ...  Major Minerals.Potassium  Major Minerals.Sodium  \
0          215       19  ...                        24                    576   
1          219       19  ...                        26                    827   
2          256       22  ...                         5                      2   
3           75       15  ...                       256                   1395   
4           94       15  ...                       136                    560   

   Major Minerals.Zinc  Vitamins.Vitamin A - IU  Vitamins.Vitamin A - RAE  \
0                 0.09                     2499                       684   
1                 0.05                     2499                       684   
2                 0.01                     3069                       840   
3                 2.66                      763                       198   
4                 2.60                     1080                       292   

   Vitamins.Vitamin B12  Vitamins.Vitamin B6  Vitamins.Vitamin C  \
0                  0.17                0.003                 0.0   
1                  0.13                0.003                 0.0   
2                  0.01                0.001                 0.0   
3                  1.22                0.166                 0.0   
4                  1.26                0.065                 0.0   

   Vitamins.Vitamin E  Vitamins.Vitamin K  
0                2.32                 7.0  
1                2.32                 7.0  
2                2.80                 8.6  
3                0.25                 2.4  
4                0.26                 2.5  

[5 rows x 48 columns]

As we can see above, each row represents a food type; the *Description* column contains a breakdown of the few ingredients that comprise the dish or food as well as details of how it was prepared. Some foods have multiple entries in the dataset because they were prepared in different ways, e.g. raw, cooked, fried, etc. Additionally, some foods have multiple entries because they were prepared with different ingredients, e.g. with or without salt, with or without sugar, etc.

The consistency of the data also presents a challenge. For example, the *Category* column contains a few entries that are not necessarily food types, e.g. 'Spices and Herbs', 'no category', etc. and occassionally foods of the same type are split into categories by brand e.g. 'Soup', 'Campbell's Soup'. I am not sure if this is intentional or not, but I will leave this data as is for now. To clean this up, I could use GPT to classify the foods into categories, but that is outside the scope of this project.

In [6]:
# Show a description of each col
df.describe()

Nutrient Data Bank Number  Alpha Carotene          Ash  Beta Carotene  \
count                7413.000000     7413.000000  7413.000000    7413.000000   
mean                14116.443680       21.210711     1.852459     159.043437   
std                  8767.416214      269.714183     2.993228    1126.285026   
min                  1001.000000        0.000000     0.000000       0.000000   
25%                  8121.000000        0.000000     0.830000       0.000000   
50%                 12539.000000        0.000000     1.240000       0.000000   
75%                 18424.000000        0.000000     2.200000       1.000000   
max                 93600.000000    14251.000000    99.800000   42891.000000   

       Beta Cryptoxanthin  Carbohydrate  Cholesterol      Choline  \
count         7413.000000   7413.000000  7413.000000  7413.000000   
mean             8.776744     21.785381    37.162822    20.673546   
std            154.184860     27.123491   119.738438    45.481990   
min              0.000000      0.000000     0.000000     0.000000   
25%              0.000000      0.490000     0.000000     0.000000   
50%              0.000000      9.290000     2.000000     0.000000   
75%              0.000000     30.590000    60.000000    20.000000   
max           7923.000000    100.000000  3100.000000  1388.000000   

             Fiber  Kilocalories  ...  Major Minerals.Potassium  \
count  7413.000000   7413.000000  ...               7413.000000   
mean      1.993147    219.655875  ...                268.348172   
std       4.292873    171.668713  ...                404.916220   
min       0.000000      0.000000  ...                  0.000000   
25%       0.000000     82.000000  ...                103.000000   
50%       0.300000    181.000000  ...                210.000000   
75%       2.300000    331.000000  ...                328.000000   
max      79.000000    902.000000  ...              16500.000000   

       Major Minerals.Sodium  Major Minerals.Zinc  Vitamins.Vitamin A - IU  \
count            7413.000000          7413.000000              7413.000000   
mean              331.590719             1.875125               767.568191   
std               977.046544             4.193682              3871.307652   
min                 0.000000             0.000000                 0.000000   
25%                31.000000             0.190000                 0.000000   
50%                86.000000             0.770000                33.000000   
75%               428.000000             2.460000               280.000000   
max             38758.000000           181.610000            100000.000000   

       Vitamins.Vitamin A - RAE  Vitamins.Vitamin B12  Vitamins.Vitamin B6  \
count               7413.000000           7413.000000          7413.000000   
mean                  99.437070              1.172903             0.269547   
std                  761.653061              4.512816             0.565116   
min                    0.000000              0.000000             0.000000   
25%                    0.000000              0.000000             0.030000   
50%                    0.000000              0.010000             0.110000   
75%                   24.000000              0.830000             0.330000   
max                30000.000000             98.890000            12.000000   

       Vitamins.Vitamin C  Vitamins.Vitamin E  Vitamins.Vitamin K  
count         7413.000000         7413.000000         7413.000000  
mean             9.075651            0.842837            9.448604  
std             63.443284            4.169756           66.067619  
min              0.000000            0.000000            0.000000  
25%              0.000000            0.000000            0.000000  
50%              0.000000            0.050000            0.000000  
75%              3.500000            0.390000            1.700000  
max           2400.000000          149.400000         1714.500000  

[8 rows x 44 columns]

Looking at the means, std. deviations, mins, and maxes, of each of our fields we can see that our data needs to be evaluated column by column to determine if each field makes sense. At face value, a mean of 21.8 for Carbohydrates and 37.2 for Cholesterol seems off, but we need to consider the units of each column. 

Of course, we will treat the Data Bank Number as a surrogate key and disregard its distribution. Similarly, Category and Description are categorical variables and we will them in our unit analysis.

Documentation for the dataset is a bit weak. For example, though the first few attributes have clearly defined units in Kaggle, the units for the remaining numeric attributes are not mentioned. Using the source of the data, the USDA FoodData Central, we can compare the values in our dataset to the source of truth and assume the units for each column. To make this simple, we will take the dataset's first entry, Butter [1001](https://fdc.nal.usda.gov/fdc-app.html#/food-details/790508/nutrients). We will assume that the units for each column are the same as the units in the source of truth unless values are off by order(s) of magnitude:

**Attribute: Unit**
1. Ash: g
2. Alpha Carotene: µg
3. Beta Carotene: µg
4. Beta Cryptoxanthin: µg
5. Carbohydrate: g
6. Cholesterol: mg
7. Choline: mg
8. Fat.Monosaturated Fat: g
9. Fat.Polysaturated Fat: g
10. Fat.Saturated Fat: g
11. Fat.Total Lipid: g
12. Fiber: g
13. Household Weights.1st Household Weight: g
14. Kilocalories: kcal
15. Lutein and Zeaxanthin: µg
16. Lycopene: µg
17. Major Minerals.Calcium: mg
18. Major Minerals.Copper: mg
19. Major Minerals.Iron: mg
20. Major Minerals.Magnesium: mg
21. Major Minerals.Phosphorus: mg
22. Major Minerals.Potassium: mg
23. Major Minerals.Sodium: mg
24. Major Minerals.Zinc: mg
25. Manganese: mg
26. Niacin: mg
27. Pantothenic Acid: mg
28. Protein: g
29. Refuse Percentage: % by volume
30. Retinol: µg
31. Riboflavin: mg
32. Selenium: µg
33. Sugar Total: g
34. Thiamin: mg
35. Vitamins.Vitamin A - RAE: µg
36. Vitamins.Vitamin B12: µg
37. Vitamins.Vitamin B6: µg
38. Vitamins.Vitamin C: µg
39. Vitamins.Vitamin E: mg
40. Vitamins.Vitamin K: µg
41. Water: g

Below, we can see the data types in this dataset are object (string), float64, and int64.

In [ ]:
# Show data types and count of attributes per that type
df.dtypes.value_counts()

As we can see below, there are only null values in the two 'Household Weight Description' columns.

In [ ]:
# Find percentage of null values and sort by missing %
df.isnull().sum().sort_values(ascending=False) / len(df) * 100

Because *2nd Household Weight* and *1st Household Weight* are the same measure expressed in different units and this column has more nulls, we will drop *2nd Household Weight* and it's corresponding *Household Weights.2nd Household Weight Description* below.

Also, because Household Weight Description doesn't really tell us anything that the *Household Weights.1st Household Weight* already containes, we will drop *Household Weights.1st Household Weight Description* as well.

In [ ]:
# Drop 'Household Weight Description' columns
df.drop(['Household Weights.2nd Household Weight Description', 'Household Weights.2nd Household Weight','Household Weights.1st Household Weight Description'], axis=1, inplace=True)

Despite having only non-null values, both categorical and numeric attributes of this dataset have a few "placeholder" values. Let's quantify these.

In [ ]:
# Find the number of values in the 'Category' column that ci match '%no category%'
df[df['Category'].str.contains('no category', case=False)]['Category'].count()/ len(df)

We can see that the *Category* column has less than 3% of its values labeled 'no category'. This is not a significant amount of data. Because we're trying to predict Kilocalories, and we don't care too much about the taxonomy of food types, we will keep these records.

Let's now target the numeric columns. We will look for values that are 0, which are likely placeholders for missing data.

In [ ]:
# Show the percentage of 0s in each numeric column of the dataset. In another table, show the percentage of 1s in each numeric column of the dataset.
df.select_dtypes(include=['float64', 'int64']).apply(lambda x: x[x == 0].count() / len(df)).sort_values(ascending=False)

Because we're dealing with a variety of foods, it makes sense that certain nutrients would be absent, so we will not impute any of these values. However, all foods should have a Weight (g) greater than 0. We will remove all rows where the Weight (g) is 0.

In [ ]:
# Drop rows where Household Weights.1st Household Weight is 0
df = df[df['Household Weights.1st Household Weight'] != 0]

Next, let's limit the scope of our dataset to exclude information that gives is found in small quantities in food. We will remove the following columns to simplify our analysis because we don't care about their correlation to KCAL:
1. **Alpha Carotene**
2. **Beta Carotene**
3. **Beta Cryptoxanthin**
4. **Lycopene**
5. **Lutein and Zeaxanthin**
6. **Retinol**
7. **Riboflavin**
8. **Selinium**
9. **Thiamin**
10. **Choline**
11. **Manganese**
12. **Niacin**
13. **Pantothenic Acid**
14. **Refuse Percentage**

In [ ]:
# Drop columns that are not relevant to our analysis
df.drop(['Alpha Carotene', 'Beta Carotene', 'Beta Cryptoxanthin', 'Lycopene', 'Lutein and Zeaxanthin', 'Retinol', 'Riboflavin', 'Selenium', 'Thiamin', 'Choline', 'Manganese', 'Niacin', 'Pantothenic Acid', 'Refuse Percentage'], axis=1, inplace=True)

We will also drop *Vitamin A - IU* because it is redundant to *Vitamin A - RAE* and the [NIH](https://ods.od.nih.gov/factsheets/VitaminA-HealthProfessional/#:~:text=The%20units%20of%20measurement%20for,beta%2Dcarotene%20%3D%200.3%20mcg%20RAE) recommends using *Vitamin A - RAE* to measure Vitamin A intake.

In [ ]:
# Drop *Vitamin A - IU* because it is redundant with *Vitamin A - RAE*
df.drop(['Vitamins.Vitamin A - IU'], axis=1, inplace=True)

## Data Profiling
Now that we’ve cleaned up the dataset, let’s see the distributions of the numeric variables. We will use a histogram to visualize the distributions of the numeric variables.

In [ ]:
# Add the numeric columns excluding Nutrient Data Bank Number to a list
numeric_cols = df.select_dtypes(include=['float64', 'int64']).drop(['Nutrient Data Bank Number'], axis=1)

#create a histogram for each numeric column
for column in numeric_cols.columns:
    sns.set(rc={"figure.figsize": (8, 4)})
    sns.displot(df[column])
    plt.show()

We can see that the distributions of the numeric variables are very right-tailed. This is expected - for example, some foods are very rich in Vitamin A like Cod Liver, while others have none.

The only variable that looks roughly normal is *Water*. Let's perform the Shapiro-Wilk test to verify if Water's or any other of the distributions are normal.

In [ ]:
# Perform the Shapiro-Wilk test on each numeric column excluding Nutrient Data Bank Number
for column in numeric_cols.columns:
    stat, p = shapiro(df[column])
    print('Statistics=%.3f, p=%.3f' % (stat, p))
    # interpret
    alpha = 0.05
    if p > alpha:
        print('{} looks Normal (fail to reject H0)'.format(column))
    else:
        print('{} does not look Normal (reject H0)'.format(column))

As we can see above, none of the distributions are normal. This is expected because the distributions are right-tailed.

In [ ]:
# N

As we can see above, none of the distributions are normal. This is expected because the distributions are right-tailed.